### PR TITLE
Refactoring location_registers and improving method to fill up category_id

### PR DESCRIPTION
**Pull request by Anntroy**

### CODE REVIEW

File: location_register/converter/koatuu.py [hunk 1]
```diff
@@ -1,148 +1,238 @@
...
 class KoatuuConverter(Converter):
 
-    #paths for the local souce file
-    LOCAL_FILE_NAME = "koatuu.json"
-    API_ADDRESS_FOR_DATASET = Register.objects.get(source_register_id="dc081fb0-f504-4696-916c-a5b24312ab6e").api_address
+    # paths for the local souce file
+    LOCAL_FILE_NAME = settings.LOCATION_KOATUU_LOCAL_FILE_NAME
+    API_ADDRESS_FOR_DATASET = Register.objects.get(
+        source_register_id=settings.LOCATION_KOATUU_SOURCE_REGISTER_ID
+    ).api_address
```
Comment (roman-tiukh) [inline location_register/converter/koatuu.py:14]: move this to `__init__`

File: location_register/converter/koatuu.py [hunk 2]
```diff
@@ -1,148 +1,238 @@
...
-                koatuu_dict[unit_name]=table_record
+                koatuu_dict[unit_name] = table_record
         return koatuu_dict
 
-    #creating a dictionary out off the values, which already exists in citydistrict table
-    def create_dictionary_for_citydistrict_table(self):#table_medel is one model from ratu_models
+    # creating a dictionary out off the values, which already exists in citydistrict table
+    def create_dictionary_for_citydistrict_table(self):# table_model is a model from ratu_models
         koatuu_dict = {}
         table_model_objects = Citydistrict.objects.all()
```
Comment (roman-tiukh) [inline location_register/converter/koatuu.py:61]: Citydistrict is bad name for model((
CityDistrict is good name.
Camel case in class names

File: location_register/converter/koatuu.py [hunk 3]
```diff
@@ -1,148 +1,238 @@
...
+    def create_district_items_list(self, object_region_name, region_dict, \
+        object_district_name, koatuu_value):
         district_items_list = []
-        if not (object_region_name in region_dict):
+        if not object_region_name in region_dict:
             return
-        region_koatuu = region_dict[object_region_name]            
-        if not (KOATUU_VALUE[:2] == str(region_koatuu.koatuu)[:2]):
+        region_koatuu = region_dict[object_region_name]
+        if not koatuu_value[:2] == str(region_koatuu.koatuu)[:2]:
```
Comment (roman-tiukh) [inline location_register/converter/koatuu.py:75]:
```suggestion
        if koatuu_value[:2] != str(region_koatuu.koatuu)[:2]:
```

File: location_register/converter/koatuu.py [hunk 4]
```diff
@@ -1,148 +1,238 @@
...
 
-    #creating list out of name and district_id from json file fot city items
-    def create_city_items_list(self, data, object_koatuu, city_dict, district_items_list, KOATUU_VALUE):
-        city_list =[]
+    # creating list out of name and district_id from json file fot city items
+    def create_city_items_list(self, city_dict, district_items_list, koatuu_value):
+        city_list = []
         for object_district_name in district_items_list:
-            if (object_district_name in city_dict) & (KOATUU_VALUE[2] =='1'):
+            if (object_district_name in city_dict) & (koatuu_value[2] == '1'):
```
Comment (roman-tiukh) [inline location_register/converter/koatuu.py:85]: what? & operator?? wrong way...
 ```suggestion
            if object_district_name in city_dict and koatuu_value[2] == '1':
```

File: location_register/converter/koatuu.py [hunk 5]
```diff
@@ -1,148 +1,238 @@
...
             return city_list
 
-    #storing data to all tables
-    def save_to_db(self,data):
+    # getting id value from category table
+    def get_category_id(self, string):
+        if string == '':
+            id_number = Category.objects.get(name='null').id
+        else:
+            id_number = Category.objects.get(name=string).id
```
Comment (roman-tiukh) [inline location_register/converter/koatuu.py:95]:
```suggestion
        id_number = Category.objects.get(name=string or 'null').id
```

Comment (roman-tiukh) [reply location_register/converter/koatuu.py:95]: what is `'null'` string?

File: location_register/converter/koatuu.py [hunk 6]
```diff
@@ -1,148 +1,238 @@
...
+        self.save_to_city_or_citydistrict(data)
+        self.writing_category_null_id(City)
+        self.writing_category_null_id(Citydistrict)
+        print("Koatuu values saved")
+
+    # writing entry to koatuu field in region table
+    def save_to_region_table(self, data):
+        for index, object_koatuu in enumerate(data):
+            region_dict = self.create_dictionary_for(Region)
+            if (object_koatuu[self.LEVEL_ONE] != '') & (object_koatuu[self.LEVEL_TWO] == ''):
```
Comment (roman-tiukh) [inline location_register/converter/koatuu.py:111]: ....
```suggestion
            if object_koatuu[self.LEVEL_ONE] and not object_koatuu[self.LEVEL_TWO]:
```

File: location_register/converter/koatuu.py [hunk 7]
```diff
@@ -1,148 +1,238 @@
...
+                object_region_name = self.get_lowercase_words_before_virgule \
+                    (object_koatuu[self.OBJECT_NAME])
```
Comment (roman-tiukh) [inline location_register/converter/koatuu.py:113]: wrong line break

File: location_register/converter/koatuu.py [hunk 8]
```diff
@@ -1,148 +1,238 @@
...
-            if (object_koatuu[self.LEVEL_ONE]!='') & (object_koatuu[self.LEVEL_TWO]!='') & (object_koatuu[self.LEVEL_THREE]!='') & (object_koatuu[self.LEVEL_FOUR]!=''):
-                object_citydistrict_name = self.get_lowercase_words_before_virgule(object_koatuu[self.OBJECT_NAME])
+                category_level_three = self.get_category_id(object_koatuu[self.CATEGORY])
+                self.save_to_city_or_citydistrict_table(
+                    object_level_three,
+                    object_city_name,
+                    district_items_list,
+                    district_dict,
+                    city_dict,
+                    category_level_three)
```
Comment (roman-tiukh) [inline location_register/converter/koatuu.py:185]:
```suggestion
                    category_level_three
                )
```

File: location_register/converter/koatuu.py [hunk 9]
```diff
@@ -1,148 +1,238 @@
...
+                self.save_to_city_or_citydistrict_table(
+                    object_level_three,
+                    object_city_name,
+                    city_items_list,
+                    city_dict,
+                    citydistrict_dict,
+                    category_level_three)
```
Comment (roman-tiukh) [inline location_register/converter/koatuu.py:192]:
```suggestion
                    category_level_three
                )
```

File: location_register/converter/koatuu.py [hunk 10]
```diff
@@ -1,148 +1,238 @@
...
+                    city_items_list,
+                    city_dict,
+                    citydistrict_dict,
+                    category_level_four)
+        print("Koatuu values to city and citydistrict saved")
 
-    #writing entry to koatuu field in city and citydistrict table
-    def save_to_city_or_citydistrict_table(self, data, object_level_number, object_level_name, level_items_list, up_level_dict, level_dict):
+    # writing entry to koatuu field in city and citydistrict table
+    def save_to_city_or_citydistrict_table(self, object_level_number, object_level_name, \
```
Comment (roman-tiukh) [inline location_register/converter/koatuu.py:216]: redundant line break orerator
```suggestion
    def save_to_city_or_citydistrict_table(self, object_level_number, object_level_name,
```

File: location_register/converter/koatuu.py (hunk 1 above)
Comment (Anntroy) [reply location_register/converter/koatuu.py:14]: Done

File: location_register/converter/koatuu.py (hunk 5 above)
Comment (Anntroy) [reply location_register/converter/koatuu.py:95]: Exists 5 types to indicate category in Koatuu register. When an object has no category, it has 'null' value.

File: location_register/converter/koatuu.py [hunk 11]
```diff
@@ -1,148 +1,238 @@
...
                 unit_name = table_record.name + str(table_record.city_id)
-                koatuu_dict[unit_name]=table_record
+                koatuu_dict[unit_name] = table_record
         return koatuu_dict
-    
 
-    #creating list out of name and district_id from json file for district items
-    def create_district_items_list(self, object_koatuu, object_region_name, region_dict, object_district_name, KOATUU_VALUE):
+    # creating list out of name and district_id from json file for district items
+    def create_district_items_list(self, object_region_name, region_dict, \
```
[the remark this answers is not in the document]
Comment (Anntroy) [reply location_register/converter/koatuu.py:69]: Ok. I'll not use it.

File: location_register/converter/koatuu.py [hunk 12]
```diff
@@ -1,148 +1,238 @@
...
+                (object_koatuu[self.LEVEL_THREE] == ''):
+                # koatuu_value is position in koatuu number wich defines district or city record
+                koatuu_value = str(object_koatuu[self.LEVEL_TWO])
+                object_district_name = self.get_lowercase_words_before_virgule \
+                    (object_koatuu[self.OBJECT_NAME])
+                district_items_list = self.create_district_items_list \
+                    (object_region_name, region_dict, object_district_name, koatuu_value)
+                category_level_two = self.get_category_id(object_koatuu[self.CATEGORY])
+                for object_district_name in district_items_list:
+                    if (object_district_name in district_dict) & (koatuu_value[2] != '1'):
```
[the remark this answers is not in the document]
Comment (Anntroy) [reply location_register/converter/koatuu.py:131]: fixed

File: location_register/converter/koatuu.py [hunk 13]
```diff
@@ -1,148 +1,238 @@
...
+                        district_koatuu = district_dict[object_district_name]
+                        district_koatuu.koatuu = object_koatuu[self.LEVEL_TWO]
+                        district_koatuu.save()
+                    elif (object_district_name in city_dict) & (koatuu_value[2] == '1'):
```
[the remark this answers is not in the document]
Comment (Anntroy) [reply location_register/converter/koatuu.py:135]: fixed

File: location_register/converter/koatuu.py [hunk 14]
```diff
@@ -1,148 +1,238 @@
...
-                KOATUU_VALUE = str(object_koatuu[self.LEVEL_TWO]) #position in koatuu wich defines district or city record 
-                object_district_name = self.get_lowercase_words_before_virgule(object_koatuu[self.OBJECT_NAME])
-                district_items_list = self.create_district_items_list(object_koatuu, object_region_name, region_dict, object_district_name, KOATUU_VALUE)
-                city_items_list = self.create_city_items_list(data, object_koatuu, city_dict_for_district_table, district_items_list, KOATUU_VALUE)
-                self.save_to_district_table(data, object_koatuu, district_dict, city_dict_for_district_table, object_region_name, region_dict, district_items_list, KOATUU_VALUE)
-            if (object_koatuu[self.LEVEL_ONE]!='') & (object_koatuu[self.LEVEL_TWO]!='') & (object_koatuu[self.LEVEL_THREE]!='') & (object_koatuu[self.LEVEL_FOUR]==''):
-                object_city_name = self.get_lowercase_words_before_virgule(object_koatuu[self.OBJECT_NAME])
+            if (object_koatuu[self.LEVEL_ONE] != '') & (object_koatuu[self.LEVEL_TWO] == ''):
```
[the remark this answers is not in the document]
Comment (Anntroy) [reply location_register/converter/koatuu.py]: fixed

File: location_register/converter/koatuu.py [hunk 15]
```diff
@@ -1,148 +1,238 @@
...
+                object_region_name = self.get_lowercase_words_before_virgule \
+                    (object_koatuu[self.OBJECT_NAME])
+            if (object_koatuu[self.LEVEL_ONE] != '') & (object_koatuu[self.LEVEL_TWO] != '') & \
+                (object_koatuu[self.LEVEL_THREE] == ''):
```
[the remark this answers is not in the document]
Comment (Anntroy) [reply location_register/converter/koatuu.py]: fixed

File: location_register/converter/koatuu.py [hunk 16]
```diff
@@ -1,148 +1,238 @@
...
+                # koatuu_value is position in koatuu number wich defines district or city record
+                koatuu_value = str(object_koatuu[self.LEVEL_TWO])
+                object_district_name = self.get_lowercase_words_before_virgule \
+                    (object_koatuu[self.OBJECT_NAME])
+                district_items_list = self.create_district_items_list \
+                    (object_region_name, region_dict, object_district_name, koatuu_value)
+                city_items_list = self.create_city_items_list \
+                    (city_dict_for_district_table, district_items_list, koatuu_value)
+            if (object_koatuu[self.LEVEL_ONE] != '') & (object_koatuu[self.LEVEL_TWO] != '') & \
+                (object_koatuu[self.LEVEL_THREE] != '') & (object_koatuu[self.LEVEL_FOUR] == ''):
```
[the remark this answers is not in the document]
Comment (Anntroy) [reply location_register/converter/koatuu.py]: fixed

File: location_register/converter/ratu.py [hunk 17]
```diff
@@ -1,11 +1,12 @@
...
 from data_ocean.models import Register
-from location_register.models.ratu_models import Region, District, City, Citydistrict, Street, Category
+from data_ocean.utils import clean_name, change_to_full_name
+from location_register.models.ratu_models import (Region, District, City, Citydistrict, Street)
 
 
 class RatuConverter(Converter):
-    API_ADDRESS_FOR_DATASET = Register.objects.get(source_register_id="a2d6c060-e7e6-4471-ac67-42cfa1742a19").api_address
+    API_ADDRESS_FOR_DATASET = Register.objects.\
+        get(source_register_id="a2d6c060-e7e6-4471-ac67-42cfa1742a19").api_address
```
[the remark this answers is not in the document]
Comment (Anntroy) [reply location_register/converter/ratu.py:9]: fixed

File: location_register/converter/koatuu.py [hunk 18]
```diff
@@ -1,148 +1,238 @@
...
+                    district_dict,
+                    city_dict,
+                    category_level_four)
+                self.save_to_city_or_citydistrict_table(
+                    object_level_four,
+                    object_citydistrict_name,
+                    city_items_list,
+                    city_dict,
+                    citydistrict_dict,
+                    category_level_four)
```
[the remark this answers is not in the document]
Comment (Anntroy) [reply location_register/converter/koatuu.py]: done

File: location_register/converter/koatuu.py [hunk 19]
```diff
@@ -1,148 +1,238 @@
...
-            level_koatuu.save(update_fields=['koatuu'])
\ No newline at end of file
+            level_koatuu.category_id = category
+            level_koatuu.save()
+
+    # writing entry to category fields where koatuu is empty
+    def writing_category_null_id(self, table_model):
+        category = self.get_category_id('null')
+        table_model_objects = table_model.objects.filter(category_id__isnull=True).count()
+        for table_record in table_model_objects:
```
[the remark this answers is not in the document]
Comment (Anntroy) [reply location_register/converter/koatuu.py]: Sorry, I've commited .count(). It was just for test

**Pull request merged**